covidthailand.py

```diff
@@ -2494,6 +2496,85 @@ def rearrange(l, *first):
 POS_AREA_SERIES = rearrange(POS_AREA_COLS, *FIRST_AREAS)
 
 
+def custom_cm(cm_name: str, size: int, last_colour: str, flip: bool = False) -> ListedColormap:
+    """Returns a ListedColorMap object built with the supplied color scheme and with the last color forced to be equal
+    to the parameter passed. The flip parameter allows to reverse the colour scheme if needed."""
+    summer = matplotlib.cm.get_cmap(cm_name)
+    if flip:
+        newcolors = summer(np.linspace(1, 0, size))
+    else:
+        newcolors = summer(np.linspace(0, 1, size))
+    newcolors[size - 1, :] = matplotlib.colors.to_rgba(last_colour)
+    return ListedColormap(newcolors)
+
+
+def cleanup_df_ages(df_all: pd.DataFrame, ages_cols: list) -> pd.DataFrame:
+    """Cleanup the data frame used for creating the 'ages' charts in case it contains missing data (NA)"""
+    df_ages = df_all['2020-12-12':][ages_cols]
+    df_ages.dropna(axis=0, how='any', inplace=True)
+
+    return df_ages
+
+
+def plot_area(df, name, prefix, title, unknown_name='Unknown', unknown_total=None, percent_fig=False, ma=False, 
+              cm='tab20'):
+    cols = [c for c in df.columns if str(c).startswith(prefix)]
+    if ma:
+        for c in cols:
+            df[f'{c} (MA)'] = df[c].rolling('7d').mean()
+        cols = [f'{c} (MA)' for c in cols]
+        ma_suffix = ' (MA)'
+    else:
+        ma_suffix = ''
+
+    if unknown_total:
+        df[f'{prefix} {unknown_name}{ma_suffix}'] = \
+            df[f'{unknown_total}{ma_suffix}'].sub(df[cols].sum(axis=1), fill_value=0).clip(lower=0)
+        cols += [f'{prefix} {unknown_name}{ma_suffix}']
+
+    if percent_fig:
+        for c in cols:
+            df[f'{c} (%)'] = df[f'{c}'] / df[cols].sum(axis=1) * 100
+        perccols = [f'{c} (%)' for c in cols]
+        
+    title = f'{title}\n'
+    title += f'Updated: {TODAY().date()}\n'
```

Comment:
I think we should be able to change this to the max date for teh first col values that isn't null?
ie when the interesting data was updated? or maybe the min date of all the cols?
thinking of the case where Cases has todays value but all the area values aren't there. The date should be yesterday in that case not today. Or in the case of testing data it should say last week if the tests haven't been updated since last week?

Reply:
Agreed.